src/parser.c

```diff
@@ -4075,11 +4092,24 @@ GumboOutput* gumbo_parse(const char* buffer) {
 GumboOutput* gumbo_parse_with_options(
     const GumboOptions* options, const char* buffer, size_t length) {
   GumboParser parser;
+  memset(&parser, 0, sizeof parser);
   parser._options = options;
+
+  if (SETJMP(parser._oom_buf)) {
+    if (parser._parser_state)
+      parser_state_destroy(&parser);
```

Comment:
Does this actually deallocate all the used memory? What about all the allocations for nodes and errors?

Reply:
It doesn’t, even if I add a call to `gumbo_destroy_output` at the end whenever there would be something non-`NULL` to pass to it.

One option that is guaranteed to work is to place all allocations in a linked list and on error destroy the list.  On the other hand, the user can always do that themselves, since Gumbo can call user-provided callbacks.

Reply:
I was considering using the [arena](https://github.com/google/gumbo-parser/pull/309) pull request to solve this problem for lua-gumbo. Unfortunately, the state of that pull request as it stands still contains dozens of unchecked calls to `malloc` and has a few other problems I can't recall off the top of my head.

Gumbo is essentially an abandoned project at this point though... It seems unlikely that any of the newer bug reports will be fixed in this repo.